### PR TITLE
Application Gateway Mutual Authentication Cmdlets.

### PR DESCRIPTION
**Pull request by HemantErappa**
## Description

Implements the cmdlets to support application gateway private link feature.

Design Review PR:
Azure/azure-powershell-cmdlet-review-pr#654

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [x] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Network/Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs [hunk 1]
```diff
@@ -91,6 +91,12 @@ public class NewAzureApplicationGatewayCommand : ApplicationGatewayBaseCmdlet
              HelpMessage = "The list of trusted root certificates")]
         public PSApplicationGatewayTrustedRootCertificate[] TrustedRootCertificate { get; set; }
 
+        [Parameter(
+             Mandatory = false,
+             ValueFromPipelineByPropertyName = true,
+             HelpMessage = "The list of trusted client CA certificate chains")]
+        public PSApplicationGatewayTrustedClientCertificate[] TrustedClientCertificates { get; set; }
```
Comment (VeryEarly) [inline src/Network/Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs:98]: please rename TrustedClientCertificates as TrustedClientCertificate,
powershell use singularity for parameter names:
https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/design-guidelines/cmdlet-best-practices.md

File: src/Network/Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs [hunk 2]
```diff
@@ -121,6 +127,12 @@ public class NewAzureApplicationGatewayCommand : ApplicationGatewayBaseCmdlet
              HelpMessage = "The list of backend http settings")]
         public PSApplicationGatewayBackendHttpSettings[] BackendHttpSettingsCollection { get; set; }
 
+        [Parameter(
+             Mandatory = false,
+             ValueFromPipelineByPropertyName = true,
+             HelpMessage = "The list of ssl profiles")]
+        public PSApplicationGatewaySslProfile[] SslProfiles { get; set; }
```
Comment (VeryEarly) [inline src/Network/Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs:134]: please rename SslProfiles as SslProfile
powershell use singularity for parameter names:
https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/design-guidelines/cmdlet-best-practices.md

File: src/Network/Network/ApplicationGateway/SslPolicy/RemoveAzureApplicationGatewaySslProfilePolicyCommand.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+// ----------------------------------------------------------------------------------
+
+using System;
+using Microsoft.Azure.Commands.Network.Models;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet("Remove", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "ApplicationGatewaySslProfilePolicy", SupportsShouldProcess = true), OutputType(typeof(PSApplicationGatewaySslProfile))]
+    public class RemoveAzureApplicationGatewaySslProfilePolicyCommand : NetworkBaseCmdlet
```
Comment (VeryEarly) [inline src/Network/Network/ApplicationGateway/SslPolicy/RemoveAzureApplicationGatewaySslProfilePolicyCommand.cs:22]: is this operation sending any request to azure?
It looks like this cmdlet only update ps object in memory. If so, no need for support should process

File: src/Network/Network/ApplicationGateway/SslPolicy/SetAzureApplicationGatewaySslProfilePolicyCommand.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Management.Automation;
+using Microsoft.Azure.Commands.Network.Models;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet("Set", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "ApplicationGatewaySslProfilePolicy", SupportsShouldProcess = true), OutputType(typeof(PSApplicationGatewaySslProfile))]
```
Comment (VeryEarly) [inline src/Network/Network/ApplicationGateway/SslPolicy/SetAzureApplicationGatewaySslProfilePolicyCommand.cs:21]: is this operation sending any request to azure?
It looks like this cmdlet only update ps object in memory. If so, no need for support should process

File: src/Network/Network/ApplicationGateway/TrustedClientCertificate/NewAzureApplicationGatewayTrustedClientCertificateCommand.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using Microsoft.Azure.Commands.Network.Models;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet("New", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "ApplicationGatewayTrustedClientCertificate", SupportsShouldProcess = true),
```
Comment (VeryEarly) [inline src/Network/Network/ApplicationGateway/TrustedClientCertificate/NewAzureApplicationGatewayTrustedClientCertificateCommand.cs:20]: is this operation sending any request to azure?
It looks like this cmdlet only update ps object in memory. If so, no need for support should process

File: src/Network/Network/ApplicationGateway/TrustedClientCertificate/RemoveAzureApplicationGatewayTrustedClientCertificateCommand.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using Microsoft.Azure.Commands.Network.Models;
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet("Remove", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "ApplicationGatewayTrustedClientCertificate", SupportsShouldProcess = true),
```
Comment (VeryEarly) [inline src/Network/Network/ApplicationGateway/TrustedClientCertificate/RemoveAzureApplicationGatewayTrustedClientCertificateCommand.cs:21]: is this operation sending any request to azure?
It looks like this cmdlet only update ps object in memory. If so, no need for support should process

File: src/Network/Network/ApplicationGateway/TrustedClientCertificate/SetAzureApplicationGatewayTrustedClientCertificateCommand.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+// ----------------------------------------------------------------------------------
+
+using Microsoft.Azure.Commands.Network.Models;
+using System;
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet("Set", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "ApplicationGatewayTrustedClientCertificate", SupportsShouldProcess = true),
```
Comment (VeryEarly) [inline src/Network/Network/ApplicationGateway/TrustedClientCertificate/SetAzureApplicationGatewayTrustedClientCertificateCommand.cs:22]: is this operation sending any request to azure?
It looks like this cmdlet only update ps object in memory. If so, no need for support should process

**Review by VeryEarly — COMMENTED** [review]
It looks to me you do not need so many cmdlet:
`New-AzApplicationGatewayClientAuthConfiguration`
`New-AzApplicationGatewaySslProfile`
`New-AzApplicationGatewayTrustedClientCertificate`

are enough, other cmdlets can be done by having a new cmdlet update-azapplicationgateway

**Review by VeryEarly — COMMENTED** [review]
can you remove localfeed and use published SDK instead?

File: src/Network/Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs (hunk 1 above)
Comment (HemantErappa) [reply src/Network/Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs:98]: This is consistent with other parameter names in  NewAzureApplicationGatewayCommand, like SslCertificates, HttpListeners, FrontendPorts, etc., which take a list of PSObjects. Also, the naming TrustedClientCertificates makes it clear to the customer that this parameter takes a list of PSObjects verses a parameter like SslPolicy which takes on a single PSObject.
https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.commands.network.newazureapplicationgatewaycommand?view=azurerm-ps

File: src/Network/Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs (hunk 2 above)
Comment (HemantErappa) [reply src/Network/Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs:134]: Same reason as above.

File: src/Network/Network/ApplicationGateway/SslPolicy/RemoveAzureApplicationGatewaySslProfilePolicyCommand.cs (hunk 3 above)
Comment (HemantErappa) [reply src/Network/Network/ApplicationGateway/SslPolicy/RemoveAzureApplicationGatewaySslProfilePolicyCommand.cs:22]: done

Comment (HemantErappa) [reply src/Network/Network/ApplicationGateway/SslPolicy/RemoveAzureApplicationGatewaySslProfilePolicyCommand.cs:22]: done

File: src/Network/Network/ApplicationGateway/SslPolicy/SetAzureApplicationGatewaySslProfilePolicyCommand.cs (hunk 4 above)
Comment (HemantErappa) [reply src/Network/Network/ApplicationGateway/SslPolicy/SetAzureApplicationGatewaySslProfilePolicyCommand.cs]: done

File: src/Network/Network/ApplicationGateway/TrustedClientCertificate/NewAzureApplicationGatewayTrustedClientCertificateCommand.cs (hunk 5 above)
Comment (HemantErappa) [reply src/Network/Network/ApplicationGateway/TrustedClientCertificate/NewAzureApplicationGatewayTrustedClientCertificateCommand.cs]: done

File: src/Network/Network/ApplicationGateway/TrustedClientCertificate/RemoveAzureApplicationGatewayTrustedClientCertificateCommand.cs (hunk 6 above)
Comment (HemantErappa) [reply src/Network/Network/ApplicationGateway/TrustedClientCertificate/RemoveAzureApplicationGatewayTrustedClientCertificateCommand.cs]: done

File: src/Network/Network/ApplicationGateway/TrustedClientCertificate/SetAzureApplicationGatewayTrustedClientCertificateCommand.cs (hunk 7 above)
Comment (HemantErappa) [reply src/Network/Network/ApplicationGateway/TrustedClientCertificate/SetAzureApplicationGatewayTrustedClientCertificateCommand.cs]: done

**Comment by HemantErappa** [comment]
> update-azapplicationgateway

**Comment by HemantErappa** [comment]
> It looks to me you do not need so many cmdlet:
> `New-AzApplicationGatewayClientAuthConfiguration`
> `New-AzApplicationGatewaySslProfile`
> `New-AzApplicationGatewayTrustedClientCertificate`
> 
> are enough, other cmdlets can be done by having a new cmdlet update-azapplicationgateway

**Comment by HemantErappa** [comment]
> It looks to me you do not need so many cmdlet:
> `New-AzApplicationGatewayClientAuthConfiguration`
> `New-AzApplicationGatewaySslProfile`
> `New-AzApplicationGatewayTrustedClientCertificate`
> 
> are enough, other cmdlets can be done by having a new cmdlet update-azapplicationgateway

ClientAuthConfiguration, SslProfile and TrustedClientCertificate also have parameters which cannot be set using set-azapplicationgateway. Please refer to Test-ApplicationGatewayCRUDWithMutualAuthentication in  src/Network/Network.Test/ScenarioTests/ApplicationGatewayTests.ps1 for examples.  Also we have these cmndlets for other objects like sslcertificate, httplistener, etc., would be better to keep it for consistency :)

File: tools/StaticAnalysis/Exceptions/Az.Network/SignatureIssues.csv [hunk 8]
```diff
@@ -372,3 +372,22 @@
...
+"Microsoft.Azure.PowerShell.Cmdlets.Network.dll","Microsoft.Azure.Commands.Network.SetAzureApplicationGatewaySslProfilePolicyCommand","Set-AzApplicationGatewaySslProfilePolicy","1","8410","Parameter DisabledSslProtocols of cmdlet Set-AzApplicationGatewaySslProfilePolicy does not follow the enforced naming convention of using a singular noun for a parameter name.","Consider using a singular noun for the parameter name."
+"Microsoft.Azure.PowerShell.Cmdlets.Network.dll","Microsoft.Azure.Commands.Network.RemoveAzureApplicationGatewaySslProfilePolicyCommand","Remove-AzApplicationGatewaySslProfilePolicy","1","8100","Remove-AzApplicationGatewaySslProfilePolicy Does not support ShouldProcess but the cmdlet verb Remove indicates that it should.	Determine if the cmdlet should implement ShouldProcess and if so determine if it should implement Force / ShouldContinue"
```
Comment (isra-fel) [inline tools/StaticAnalysis/Exceptions/Az.Network/SignatureIssues.csv:388]: This line and the followings were strangely formatted, causing static analysis tool failing to parse the file. Please fix them.

```suggestion
"Microsoft.Azure.PowerShell.Cmdlets.Network.dll","Microsoft.Azure.Commands.Network.RemoveAzureApplicationGatewaySslProfilePolicyCommand","Remove-AzApplicationGatewaySslProfilePolicy","1","8100","Remove-AzApplicationGatewaySslProfilePolicy Does not support ShouldProcess but the cmdlet verb Remove indicates that it should.","Determine if the cmdlet should implement ShouldProcess and if so determine if it should implement Force / ShouldContinue"
```

File: src/Network/Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs (hunk 1 above)
Comment (isra-fel) [reply src/Network/Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs:98]: Yes, this parameter takes a list of objects, but it _also_ takes one signle object, aka `-TrustedClientCertificates @{...}` is totally OK. That is why we ask everyone to use singular name.

Considering exisiting parameters I won't request you to revise, but I do hope you understand why.

**Review by isra-fel — CHANGES_REQUESTED** [review]

**Review by isra-fel — APPROVED** [review]

**Pull request merged**